cupy/core/_routines_math.pyx

```diff
@@ -341,6 +342,109 @@ def _inclusive_scan_kernel(src_dtype, dtype, block_size, op, src_c_cont,
     return module.get_function(name)
 
 
+@_util.memoize(for_each_device=True)
+def _inclusive_scan_kernel_shfl(src_dtype, dtype, block_size, op, src_c_cont,
+                                out_c_cont):
+    """return Prefix Sum(Scan) cuda kernel using warp shuffle
+
+    e.g
+    if blocksize * 2 >= len(src)
+    src [1, 2, 3, 4]
+    dst [1, 3, 6, 10]
+
+    if blocksize * 2 < len(src)
+    block_size: 2
+    src [1, 2, 3, 4, 5, 6]
+    dst [1, 3, 6, 10, 5, 11]
+
+    Args:
+        dtype: src, dst array type
+        block_size: block_size
+
+    Returns:
+         cupy.cuda.Function: cuda function
+    """
+
+    name = 'inclusive_scan_kernel_shfl'
+    src_dtype = get_typename(src_dtype)
+    dtype = get_typename(dtype)
+    op_char = {scan_op.SCAN_SUM: '+', scan_op.SCAN_PROD: '*'}
+    identity = {scan_op.SCAN_SUM: 0, scan_op.SCAN_PROD: 1}
+    source = string.Template("""
+    extern "C" __global__ void ${name}(
+        const CArray<${src_dtype}, 1, ${src_c_cont}> src,
+        CArray<${dtype}, 1, ${out_c_cont}> dst){
+        long long n = src.size();
+        __shared__ ${dtype} temp[${block_size} / 32];
+        unsigned int warp_id = threadIdx.x / 32;
+        unsigned int lane_id = threadIdx.x % 32;
+        unsigned int idx1 = 2 * (threadIdx.x + blockIdx.x * blockDim.x);
+        unsigned int idx2 = idx1 + 1;
+        ${dtype} v1, v2;
+        v1 = (idx1 < n) ? (${dtype})src[idx1] : (${dtype})${identity};
+        v2 = (idx2 < n) ? (${dtype})src[idx2] : (${dtype})${identity};
+
+        v2 ${op}= v1;
+        for (int i = 1; i < 32; i <<= 1) {
```

Comment:
Can this loop be unrolled?

Reply:
Yes, this loop is unrollable, but since this kernel is memory bandwidth bound, unrolling it has no impact on performance.